Use API Policy; Restrict AuthClientPolicy to `POST /api/groups`

lyzadanger opened the pull request:
~~~DEPENDS ON #5223~~~ <-- That's been merged and this has been rebased

This PR enables the new `APIAuthenticationPolicy`, which is a no-op for all API endpoints (still uses Token Auth for them) except `POST /api/groups`. Thus, a private group may be created by an auth client with a forwarded user.

At this stage, there is a check in the view itself to validate that there is a `request.user` present (versus an auth_client that passed authentication but doesn't have a forwarded user to work with).

lyzadanger commented:
> I started to run into problems, though, when I started trying to call the group create API using a client_id, client_secret and X-Forwarded-User. If I try to call it with an invalid userid it crashes. Here "seanh@partner.org" is invalid (doesn't start with "acct:")

This is fixed

lyzadanger commented:
> If I do the same but with an X-Forwarded-User that is valid but doesn't exist I get a 404.

Unfortunately beyond the scope of this PR, but part of the broader auth'n and auth'z fixes: this is the behavior of all of our API endpoints that use a `security.Authenticated` predicate—if that fails, Pyramid raises a 404, not a 401. It's on my list, but not really addressable here.

FWIW this is already the case if you hit this endpoint un-auth'd.

lyzadanger commented:
> >>> requests.post("http://localhost:5000/api/groups", auth=("587db8f4-495d-11e8-a744-dbd946aa49e9", "wzbQfHULj2tymqHF6bSQUkq4GNDJpwZVpgDmVX6o5qY"), headers={"X-Forwarded-User": "acct:seanh@partner.org"}, data=json.dumps(dict(name="foobar"))).content
ValueError: Organization's authority hypothes.is must match the group creator's authority partner.org.

Dear deity; this sure caused some fuss. This PR is now dependent on #5223

lyzadanger commented:
> I created a client_credentials auth client for the hypothes.is authority to see if it would work. As long as the userid I give in the X-Forwarded-User is an existing valid hypothes.is-authority userid then it does work, but I do get a unicode warning printed in the terminal:

I can't reproduce this no way no how :). Help?

seanh commented:
For reference, here's an incomplete diagram of h's auth setup (as it exists on this branch) that I found useful in reviewing this. My its grown!

![2018-08-24 17 58 27](https://user-images.githubusercontent.com/22498/44597483-ab5b0680-a7c7-11e8-9f63-a735d65968b7.jpg)

The way all the classes get plugged together in [`h/auth/__init__.py`](https://github.com/hypothesis/h/blob/9e4ae4f43c5e93e6a875fa473966859fceda3ada/h/auth/__init__.py) is dizzying. I think the `CONSTANTS` used in the file (most of them used only in one place, some of them not actually constants because they get conditionally reassigned) are hindering comprehension (increasing travel needed to follow the code, etc) more than helping.

seanh commented:
I've begun putting together a list of cases for myself to manually test on this branch:

- [ ] An AuthClient (client_id and client_secret)-authorized request to anything other than `POST /api/groups` should fail
- [ ] A user `Bearer` token in the `Authorization` header should override a `client_id` and `client_secret` in the HTTP Basic Auth
- [ ] What happens when a request has both a `Bearer` token in the `Authorization` header and a `client_id` and `client_secret` in the HTTP Basic Auth, and the `Bearer` token is invalid?
- [ ] And AuthClient-authorized post to `/api/groups` should fail if there's no `X-Forwarded-User`
- [ ] Test successfully creating a private group with an AuthClient-authorized, `X-Forwarded-User` post to `/api/groups`
- [ ] It should still be possible to use the _user_ create API with a `client_id` and `client_secret` but no `X-Forwarded-User`
- [ ] What happens if you call _user_ create API with a `client_id` and `client_secret` _and_ an `X-Forwarded-User`?
- [ ] It should still be possible to call the _annotation_ create API using a developer token in a Bearer header
- [ ] It should still be possible to call the _group_ create API using a developer token in a Bearer header
- [ ] Try invalid `client_id` and/or `client_secret` in the HTTP basic auth
- [ ] Try a valid `client_id` and `client_secret` but with an invalid userid in the `X-Forwarded-User` header
- [ ] Try a valid `client_id` and `client_secret` with a valid but doesn't exist userid in the `X-Forwarded-User` header
- [ ] ...

seanh commented:
I've rebased this branch on top of #5223 locally but I'm out of time for this week, will have to finish testing and reviewing this next week

seanh commented on h/auth/policy.py:
```diff
@@ -326,5 +337,19 @@ def _is_api_request(request):
+    This is intended to be temporary.
+    """
+    return (request.path.startswith('/api/groups') and
+            request.method == 'POST')
```
Assuming it'd be temporary is probably how the other hard-coded paths that're in here got in, and they're still here!

seanh commented:
One thing I'm noticing (for the future): Maybe the API should check for the presence of each different possible form of authentication (bearer token, basic auth, ...) and always fail if more than one form of authentication is present in the request. That would simplify the behaviour, as the API would no longer have behaviours around which form of auth overrides the other when more than one is provided, and what if two forms are provided but one is invalid? Etc.

seanh commented:
Oh, and I'm still seeing SQLAlchemy warnings about unicode strings, but I don't have instructions for how to reproduce those yet. Lets fix the group member add problem, and then when I re-test I'll figure that out too?

lyzadanger commented:
@seanh I've fixed the idiocy and I've refactored tests to parametrize them for paths right now. It's a bit verbose but I didn't want to try to get too clever at this stage in the game. Please re-review!

seanh commented:
I've opened an issue for a group member add API crash that's reproducible on master: https://github.com/hypothesis/h/issues/5246

seanh commented on h/auth/policy.py:
```diff
@@ -326,5 +348,19 @@ def _is_api_request(request):
+    This is intended to be temporary.
+    """
+    return (re.match(AUTH_TOKEN_PATH_PATTERN, request.path) and
+            request.method == 'POST')
```
Can this be `request.matched_route.name == "api.groups" and request.method == "POST"` instead? That's the correct way to do it. The `/api/groups` path is served by the `"api.groups"` route:

https://github.com/hypothesis/h/blob/673189bb14e1151c5d5de77dce85ee8a067eeecf/h/routes.py#L110

And that route is only used [by the group read API](https://github.com/hypothesis/h/blob/673189bb14e1151c5d5de77dce85ee8a067eeecf/h/views/api/groups.py#L16-L19) (with `request_method='GET'`) and [by the group create API](https://github.com/hypothesis/h/blob/673189bb14e1151c5d5de77dce85ee8a067eeecf/h/views/api/groups.py#L39-L42) with `request_method='POST'`. This won't break if the API paths change but mostly it's simpler: no string pattern matching mistakes, no regex needed.

The group member APIs use a different route name, `"api.group_member"`, so they won't match:

https://github.com/hypothesis/h/blob/673189bb14e1151c5d5de77dce85ee8a067eeecf/h/routes.py#L113-L116

seanh replied:
In the tests [`pyramid_request.matched_route` is a mock](https://github.com/hypothesis/h/blob/673189bb14e1151c5d5de77dce85ee8a067eeecf/tests/h/conftest.py#L242) so test setup will have to do `pyramid_request.matched_route.name = "whatever"` to test it against different `matched_route` values.

seanh commented on tests/h/auth/policy_test.py:
```diff
@@ -31,6 +31,19 @@
+    '/api/badge',
+    '/api',
+    '/groups/'
+)
```
Could these constants be moved into the `TestAPIAuthenticationPolicy` that they're used in? Just to reduce the amount of travel needed to understand the code. Although if `request.matched_route` is used the code'll be doing simple `==` against a single possible matching string value, instead of any kind of pattern match, so testing with multiple correct and incorrect strings won't be necessary anymore anyway.
